### PR TITLE
Changes to HTTPMetricHandler to fix Issue #732

### PR DESCRIPTION
**Pull request by dhoard**
Changes to HTTPServer and HTTPMetricHandler to resolve getting Content-Length=0 when Transfer-Encoding=chunked. Refactored TestHTTPTest for cleaner testing.

Signed-off-by: Doug Hoard <doug.hoard@gmail.com>

### CODE REVIEW

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java [hunk 1]
```diff
@@ -115,316 +93,412 @@ public void init() throws IOException {
...
+      HttpResponse httpResponse = httpRequest.execute();
+
+      Assert.assertTrue(
+              (httpResponse.getHeader("content-length") == null)
+                      ^ (httpResponse.getHeader("transfer-encoding") == null));
+
+      String body = httpResponse.getBody();
+      assertThat(body).doesNotContain("a 0.0");
+      assertThat(body).doesNotContain("b 0.0");
+      assertThat(body).doesNotContain("c 0.0");
```
Comment (fstab) [inline simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java:252]: Can be replaced with `assertThat(body).isEmpty()`.

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java [hunk 2]
```diff
@@ -550,11 +680,29 @@ public static SSLContext createSSLContext(String sslContextType, String keyStore
...
+  }
+
+  /**
+   * Creates an HttpsConfiguration
+   * 
    * @param sslContext
    * @return HttpsConfigurator
    */
-  private static HttpsConfigurator createHttpsConfigurator(SSLContext sslContext) {
+  private final static HttpsConfigurator createHttpsConfigurator(SSLContext sslContext) {
```
Comment (fstab) [inline simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java:705]: `final` should be removed

**Review by fstab — CHANGES_REQUESTED** [review]
Thaks a lot for the PR! I like the idea of refactoring the `request()` methods to make the test clearer.

However, it would be good if after the refactoring the test had less lines of code, because the new `HttpRequest` / `HttpResponse` classes allowed expressing the test more concisely. Currently, the test grew from 576 lines to 724 lines even though all the `request()` methods are removed. With a bit of refactoring we could reduce the boiler plate:

```java
Assert.assertTrue(
        (httpResponse.getHeader("content-length") == null)
        ^ (httpResponse.getHeader("transfer-encoding") == null));
```

This is copied 16 times. It would be good to have one test to make sure that the `content-length` and `transfer-encoding` headers are correct, and not to copy-and-paste this to all tests.


```java
HttpRequest httpRequest = new HttpRequest.Builder()
        .withURL("http://localhost:" + httpServer.getPort() + "/metrics?x")
        .build();

HttpResponse httpResponse = httpRequest.execute();
String body = httpRequest.execute().getBody();
```

The relevant part here is the `?x` URL parameter. This was easier to see in the previous single line `String body = request(s, "?x").body;`. Maybe we can add a helper method `request(HTTPServer, String)` that returns the initialized `Builder`? Then we could cover the simple case with `request(httpServer, "?x").build().execute().getBody()`. To reduce this even further, we could just move the `Builder` methods directly into the `HttpRequest`, then you could write `request(httpServer, "?x").execute().getBody()`.

For more complex tests we could still write something like

```
request(httpServer, "?x")
    .withHeader(...)
    .withAuthorization(...)
    .execute().getBody();
```

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java (hunk 1 above)
Comment (dhoard) [reply simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java:252]: Agreed.

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java (hunk 2 above)
Comment (dhoard) [reply simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java:705]: Agreed.

**Comment by dhoard** [comment]
@fstab I have refactored the code to _mostly_ align with your suggestions.

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/HttpResponse.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                this.headers.put(header.toLowerCase(), values);
+            }
+        }
+
+        // HTTP 1.0 and 1.1 allow both headers to be null if "Connection: close"
+        if ((getHeader("content-length") != null) || (getHeader("transfer-encoding") != null)) {
+            if (false == ((getHeader("content-length") == null) ^ (getHeader("transfer-encoding") == null))) {
+                throw new IOException("Invalid HTTP response, should only contain Connect-Length or Transfer-Encoding");
+            }
+        }
```
Comment (fstab) [inline simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/HttpResponse.java:46]: Can be simplified to
```java
if (getHeader("content-length") != null && getHeader("transfer-encoding") != null) {
    throw new IOException("Invalid HTTP response, should only contain Connect-Length or Transfer-Encoding");
}
```

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java [hunk 4]
```diff
@@ -115,316 +107,269 @@ public void init() throws IOException {
...
+
     try {
-      String response = request(s, "?name[]=a&name[]=b").body;
-      assertThat(response).doesNotContain("a 0.0");
-      assertThat(response).contains("b 0.0");
-      assertThat(response).doesNotContain("c 0.0");
+      String body = createHttpRequestBuilder(httpServer, "/metrics?name[]=a&name[]=b").build().execute().getBody();
+      assertThat(body).doesNotContain("a 0.0");
+      assertThat(body).doesNotContain("b 0.0");
+      assertThat(body).doesNotContain("c 0.0");
```
Comment (fstab) [inline simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java:206]: Can be simplified to
```java
assertThat(httpResponse.getBody()).isEmpty();
```

**Review by fstab — CHANGES_REQUESTED** [review]
It's looking great, thanks a lot! I added two little nitpick remarks, apart from that it's ready to be merged.

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java (hunk 4 above)
Comment (dhoard) [reply simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/TestHTTPServer.java:206]: Agreed. Simplified the code to

```
HttpResponse httpResponse = createHttpRequestBuilder(httpServer, "/metrics?name[]=a&name[]=b").build().execute();
assertThat(httpResponse.getBody()).isEmpty();
```

File: simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/HttpResponse.java (hunk 3 above)
Comment (dhoard) [reply simpleclient_httpserver/src/test/java/io/prometheus/client/exporter/HttpResponse.java:46]: Agreed.

**Comment by fstab** [comment]
Forgot to push?

**Comment by dhoard** [comment]
> Forgot to push?

Yep. pushed. Appologies.

**Comment by fstab** [comment]
Thanks a lot!

**Pull request merged**